pkg/util/aad/aad.go

```diff
@@ -39,12 +39,11 @@ func GetToken(ctx context.Context, log *logrus.Entry, clientID, clientSecret, te
 	// during credentials rotation this can take time to propagate
 	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
 	defer cancel()
-
+	done := false
 	// NOTE: Do not override err with the error returned by
 	// wait.PollImmediateUntil. Doing this will not propagate the latest error
 	// to the user in case when wait exceeds the timeout
 	_ = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
```

Comment:
I still think that this function should not be anonymous

Reply:
@facchettos : we haven't touched that part for the new changes, this is something already exists.

Reply:
yes you have. you have take the `done` variable out of the function and this is dangerous and makes the code harder to understand.

Reply:
The `_ = wait.PollImmediateUntil(10*time.Second, func() (bool, error)` can be changed to
`err = wait.PollImmediateUntil(10*time.Second, func() (bool, error)` and err can be used instead.

If you look how the poller works, it propagetes the error from internal function. It also add some timeout errors or context errors, but is what you want to use when propagating it out.

Reply:
@petrkotas : I guess this change is out of scope of the ADO card related to this PR, also check the comments in before this anonymous func:

>   // NOTE: Do not override err with the error returned by
> // wait.PollImmediateUntil. Doing this will not propagate the latest error
> // to the user in case when wait exceeds the timeout 

@facchettos : without taking that done variable out, if you have any suggestion to solve this, please let us know

Reply:
```go
        done := false 
	_ = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
		authorizer.RefreshAndParse(&done)
	}, timeoutCtx.Done())



func (authorizer authorizer) RefreshAndParse(done *bool ) (bool, error){

		*done, err = authorizer.RefreshWithContext(ctx, log)
		if err != nil {
			err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalCredentials, "properties.servicePrincipalProfile", "The provided service principal credentials are invalid.")
		}
		if !*done || err != nil {
			return false, err
		}

		p := &jwt.Parser{}
		claims := jwt.MapClaims{}
		_, _, err = p.ParseUnverified(authorizer.OAuthToken(), claims)
		if err != nil {
			err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", "The provided service principal generated an invalid token.")
			return false, err
		}

		for _, claim := range []string{"altsecid", "oid", "puid"} {
			if _, found := claims[claim]; found {
				return true, nil
			}
		}

		// populate err with a user-facing error that will be visible if we're
		// not successful.
		log.Info("token does not contain the required claims")
		err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalClaims, "properties.servicePrincipalProfile", "The provided service principal does not give an access token with at least one of the claims 'altsecid', 'oid' or 'puid'.")

		return false, nil
}
```

now you have something that is testable instead of having anonymous structs that are absolutely unseparable from the code. you could also mock that to test the bigger function.